Modules/_posixsubprocess.c

```diff
@@ -1050,50 +1049,34 @@ subprocess_fork_exec(PyObject* self, PyObject *args)
     }
 #endif
 
-    Py_XDECREF(cwd_obj2);
-
     if (need_after_fork)
         PyOS_AfterFork_Parent();
-    if (envp)
-        _Py_FreeCharPArray(envp);
-    if (argv)
-        _Py_FreeCharPArray(argv);
-    _Py_FreeCharPArray(exec_array);
-
-    /* Reenable gc in the parent process (or if fork failed). */
-    if (_enable_gc(need_to_reenable_gc, gc_module)) {
-        pid = -1;
-    }
-    PyMem_RawFree(groups);
-    Py_XDECREF(preexec_fn_args_tuple);
-    Py_XDECREF(gc_module);
 
-    if (pid == -1) {
+cleanup:
+    if (saved_errno != 0 || pid == -1) {
```

Comment:
I added a pid == -1 check here as a defensive coding measure as that is what we use below to exit the function with a NULL return.  in theory it should always be true if we trust that fork() or vfork() behave as documented.  but if for some reason they don't, we still have a bad pid and need to return an error.  it'll be an odd `OSError: Success` but that counts and is enough to suggest something strange is up on the system.

Reply:
based on the CI failures this was apparently wrong to do?  i'm still staring at the code to understand why.

Reply:
okay, reverting this wrong edit I made.  pid == -1 being the default means all of the pre-existing error paths that are not errno based have a PyErr already set and need to exit without triggering this.

i don't like that we'd miss fork returning -1 with zero errno but that likely is not a realistic worry.

Reply:
Yes, I pondered "`do_fork_exec()` fails => `saved_errno != 0`" implication too when I was writing the PR. I think this is not a realistic worry too, but if we want, we can make `pid` zero by default since it's an invalid PID as well (and change the return statement of `subprocess_fork_exec()` accordingly).

Anyway, thanks for the review and the fixup!